Batch Changes - update vision, strategy, plan page! (#4562)
sourcegraph/about: handbook/engineering/batch-changes/direction.md

# Batch Changes direction

Batch Changes is a tool to find code that needs to be changed and change it at scale by running code. This page outlines the vision, strategy and goals of the Batch Changes team.

#### Quick links

- [Planning board](https://github.com/orgs/sourcegraph/projects/119)
- [Demo video](https://www.youtube.com/watch?v=eOmiyXIWTCw)
- [Batch Changes documentation](https://docs.sourcegraph.com/batch_changes)

# Vision

#### 1 Year vision

> > Enterprise developers can easily create changesets across many repositories and codehosts, and track them to completion.

Reviewer: Sorry, pet peeve (and a hill I'm prepared to die on): let's not use quoting to higlighting something. Make it bold, italic, give it a separate style, but not the quote style.

> > Enterprise **developers** can easily create changesets across many repositories and codehosts, and track them to completion.

Reviewer 2: ```suggestion
> Enterprise developers can easily create changesets across many repositories and codehosts, and track them to completion.
```

#### 3 Year vision

> Developers can easily create changesets across many repositories and codehosts, and track them to completion.

#### 10 year vision

> Automated code changes are a common practice for all developers

## Mission & Strategy

### Mission

> We allow developers to focus on changing their code, without having to do any plumbing. We help them drive changesets to completion without having to rely on spreadhseets and painful coordination. Later on, we want to make writing and running code that changes code easier to learn and use.

Author: Thank you! That was very unclear, used your wording instead.

### Strategy

To deliver our vision, we will:

- First, get adopted by the Sourcegraph customer base, including large enterprises. We will target customers with advanced practices and workflow, and collaborate with them to build a category-defining product that all other companies will use.

> - Concurrently, spread usage (MAUs) within our customer, and make sure we deliver on our success metric (changesets merged)

Reviewer 2: ```suggestion
- Concurrently, spread usage (MAUs) within our customer base, and make sure we deliver on our success metric (changesets merged).
```

- Then, create low-effort entrypoints into Batch Changes, and allow users unfamiliar with the practice of automating code changes to onboard with a smooth lerning curve. Make onboarding Batch Changes self-service for most users, to help our customers be successfull and prepare for launching on Cloud.
- Then, release Batch Changes on Sourcegraph Cloud, and test and learn how we can serve individuals and smaller teams

> - Eventually, address the problem end to end by becoming the go-to place for code change tools and recipes.

Author: I'll add an indication that we are at the first two bullet points

# Guiding Principles

## Themes

To deliver the strategy, here are the themes we want to focus on for the next 3 years:

- [Enterprise scale](#enterprise-scale)
- [Discovery and onboarding](#discovery-and-onboarding)
- [Adoption through low-effort usage patterns](#adoption-through-low-effort-usage-patterns)
- [Cloud](#cloud)
- [Automate common code changes](#automate-common-code-changes)

### Enterprise scale

Batch Changes is succesfully adopted by very large companies, with 10,000s of repositories. While there is no technical limitation for Batch Changes to run at that scale, there are 2 limitations in practice:

- creating changesesets locally takes too long to be practical for very large scale changes
- in companies with 10,000s of repositories, it is difficult for batch changes maintainers to understand what repositories they should target with a change, and to analyse progress of a batch change.

To drive enterprise success and stickiness, we need to:

> - deliver a great experience in creating changesets at enterprise scale (10,000s reposiories), resulting in more users onboarding batch changes.

Author: Done

> - allow users to find out where to create batch changes and and understand what to do to get changesets merged in batch changes with 1,000 of changesets. This will increase the number of changesets merged, our key success metric.

Reviewer 2: ```suggestion
- allow users to find out where to create batch changes and and understand what to do to get changesets merged in batch changes with thousands of changesets. This will increase the number of changesets merged, our key success metric.
```
